### PR TITLE
[TG-1793] Complete instanceof for Java.

### PR DESCRIPTION
**Pull request by mgudemann**
The current implementation did not consider correctly the case of `null` in
`ID_java_instanceof`. An instanceof with `null` always fails. For `a instanceof
B` this now adds `a != null` to the condition.

### CODE REVIEW

**Comment by tautschnig** [comment]
This is a bug fix, please add a corresponding regression test.

File: src/goto-programs/remove_instanceof.cpp [hunk 1]
```diff
@@ -113,15 +113,20 @@ std::size_t remove_instanceoft::lower_instanceof(
...
   exprt::operandst or_ops;
   for(const auto &clsname : children)
   {
     constant_exprt clsexpr(clsname, string_typet());
     equal_exprt test(newsym.symbol_expr(), clsexpr);
     or_ops.push_back(test);
   }
-  expr=disjunction(or_ops);
+  exprt non_null_expr = notequal_exprt(
+    check_ptr, null_pointer_exprt(to_pointer_type(check_ptr.type())));
```
Comment (tautschnig) [inline src/goto-programs/remove_instanceof.cpp]: I'd suggest `notequal_exprt non_null_expr( ... )`

**Review by tautschnig — CHANGES_REQUESTED** [review]
See my earlier comment: needs a regression test.

File: src/goto-programs/remove_instanceof.cpp [hunk 2]
```diff
@@ -113,15 +113,20 @@ std::size_t remove_instanceoft::lower_instanceof(
...
+  // (http://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.20.2)
   exprt::operandst or_ops;
   for(const auto &clsname : children)
   {
     constant_exprt clsexpr(clsname, string_typet());
     equal_exprt test(newsym.symbol_expr(), clsexpr);
     or_ops.push_back(test);
   }
-  expr=disjunction(or_ops);
+  exprt non_null_expr = notequal_exprt(
```
Comment (thk123) [inline src/goto-programs/remove_instanceof.cpp]: I would be tempted to put the construction of the non_null_expr directly below the comment.

The LHS can be more precisely typed as a `notequal_exprt`.

**Review by thk123 — APPROVED** [review]
I know we have a test in TG, but if possible might be good to have a regression in JBMC too.

File: src/goto-programs/remove_instanceof.cpp [hunk 3]
```diff
@@ -113,15 +113,20 @@ std::size_t remove_instanceoft::lower_instanceof(
   newinst->source_location=this_inst->source_location;
   newinst->function=this_inst->function;
 
-  // Replace the instanceof construct with a big-or.
+  // Replace the instanceof construct with a conjunction of non-null and the
+  // disjunction of all possible object types. According to the Java
+  // specification, null instanceof T is false for all possible values of T.
+  // (http://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.20.2)
+  notequal_exprt non_null_expr = notequal_exprt(
```
Comment (tautschnig) [inline src/goto-programs/remove_instanceof.cpp]: Is the use of the assignment a left-over?

**Comment by mgudemann** [comment]
yes, is removed now, thanks!

**Comment by tautschnig** [comment]
The changes now look fine, but I'd really really appreciate a test here.

**Comment by mgudemann** [comment]
@tautschnig @thk123 I added a public regression test for this

**Review by tautschnig — APPROVED** [review]

**Review by thk123 — APPROVED** [review]

**Pull request merged**